test/jdk/sun/security/krb5/native/NativeCacheTest.java

```diff
@@ -0,0 +1,212 @@
+ *   --add-exports java.security.jgss/sun.security.jgss.krb5=ALL-UNNAMED
+ *   --add-exports java.base/sun.security.util=ALL-UNNAMED
+ *   --add-exports java.base/jdk.internal.misc=ALL-UNNAMED
+ *   --add-opens java.security.jgss/sun.security.krb5=ALL-UNNAMED
+ *   --add-opens java.security.jgss/sun.security.krb5.internal=ALL-UNNAMED
+ *   --add-opens java.base/sun.security.util=ALL-UNNAMED
+ *   --enable-native-access=ALL-UNNAMED
+ *   -Djdk.net.hosts.file=TestHosts
+ *   NativeCacheTest
+ */
+
+import sun.security.krb5.Credentials;
+import javax.security.auth.login.LoginContext;
+import javax.security.auth.Subject;
+import javax.security.auth.kerberos.KerberosTicket;
+import java.io.File;
+
+/**
+ * Test JAAS access to in-memory credential caches.
+ *
+ * This test validates that JAAS can access in-memory credential caches
+ * on Linux through the native enhancement, using TGTs from OneKDC.
+ */
+public class NativeCacheTest {
+
+    public static void main(String[] args) throws Exception {
+        try {
+            // Create TGT using OneKDC (in file cache)
+            createTGTWithOneKDC();
+
+            // Copy TGT to in-memory cache using JNI
+            String inMemoryCacheName = copyTGTToInMemoryCache();
+
+            // Test JAAS access to in-memory cache
+            testJAASAccessToInMemoryCache(inMemoryCacheName);
+
+        } catch (UnsatisfiedLinkError e) {
+            System.out.println("Kerberos native library not available - test skipped");
+            return;
+        } catch (Exception e) {
+            System.err.println("Test failed: " + e.getMessage());
+            throw e;
+        }
+    }
+
+    /**
+     * Use OneKDC to create a TGT via the JAAS LoginModule
+     */
+    private static void createTGTWithOneKDC() throws Exception {
+        System.out.println("Creating TGT via OneKDC");
+
+        OneKDC kdc = new OneKDC(null);
+        kdc.writeJAASConf();
+
+        // Force JAAS to save credentials to file cache for copying
+        System.setProperty("test.kdc.save.ccache", "onekdc_cache.ccache");
+
+        try {
+            // Authenticate using the JAAS LoginModule
+            LoginContext lc = new LoginContext("com.sun.security.jgss.krb5.initiate",
+                                               new OneKDC.CallbackForClient());
+            lc.login();
+
+            // Verify authentication
+            Subject subject = lc.getSubject();
+            KerberosTicket ticket = subject.getPrivateCredentials(KerberosTicket.class).iterator().next();
+
+            System.out.println("JAAS authentication successful");
+            System.out.println("TGT: " + ticket.getClient() + " -> " + ticket.getServer());
+
+        } catch (Exception e) {
+            System.out.println("JAAS authentication failed: " + e.getMessage());
+        }
+    }
+
+    /**
+     * Copy the TGT to an in-memory cache using JNI
+     */
+    private static String copyTGTToInMemoryCache() throws Exception {
```

Comment:
All native methods are called here with the same order. Is it possible to just create one single native method and do everything inside?